File: analytics/src/test/java/io/appmetrica/analytics/impl/crash/ReadAndReportRunnableTest.java
package io.appmetrica.analytics.impl.crash;

import io.appmetrica.analytics.coreapi.internal.backport.Consumer;
import io.appmetrica.analytics.coreapi.internal.backport.Function;
import io.appmetrica.analytics.impl.EventsManager;
import io.appmetrica.analytics.impl.TestsData;
import io.appmetrica.analytics.impl.client.ClientConfiguration;
import io.appmetrica.analytics.impl.client.ClientConfigurationTestUtils;
import io.appmetrica.analytics.impl.crash.jvm.JvmCrash;
import io.appmetrica.analytics.logger.appmetrica.internal.PublicLogger;
import io.appmetrica.analytics.testutils.CommonTest;
import java.io.File;
import java.util.HashMap;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ReadAndReportRunnableTest extends CommonTest {

    @Mock
    private File crashFile;
    @Mock
    private Function<File, JvmCrash> fileReader;
    @Mock
    private Consumer<File> finalizator;
    @Mock
    private Consumer<JvmCrash> crashConsumer;
    private ReadAndReportRunnable<JvmCrash> readAndReportRunnable;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        doReturn(true).when(crashFile).exists();
        readAndReportRunnable = new ReadAndReportRunnable<JvmCrash>(crashFile, fileReader, finalizator, crashConsumer);
    }

    @Test
    public void testEmpty() {
        doReturn(null).when(fileReader).apply(any(File.class));
        readAndReportRunnable.run();
        verify(fileReader).apply(crashFile);
        verify(crashConsumer, never()).consume(any(JvmCrash.class));
    }

    @Test
    public void testCrashHandled() {
        ClientConfiguration stubbedConfiguration = ClientConfigurationTestUtils.createStubbedConfiguration();
        stubbedConfiguration.getReporterConfiguration().setApiKey(TestsData.generateApiKey());
        final JvmCrash crash = new JvmCrash(
                EventsManager.unhandledExceptionReportEntry("crash name", "crash value".getBytes(), mock(PublicLogger.class)),
                stubbedConfiguration,
                mock(HashMap.class)
        );
        doReturn(crash).when(fileReader).apply(any(File.class));
        readAndReportRunnable.run();
        ArgumentCaptor<JvmCrash> newCrash = ArgumentCaptor.forClass(JvmCrash.class);
        verify(crashConsumer).consume(newCrash.capture());
        assertThat(crash).usingRecursiveComparison().isEqualTo(newCrash.getValue());
    }

    @Test
    public void testCrashFileDeleted() {
        when(crashFile.exists()).thenReturn(true);
        readAndReportRunnable.run();
        verify(finalizator).consume(crashFile);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/crash/ReadAndReportRunnableTest.kt
package io.appmetrica.analytics.impl.crash

import io.appmetrica.analytics.coreapi.internal.backport.Consumer
import io.appmetrica.analytics.coreapi.internal.backport.Function
import io.appmetrica.analytics.impl.crash.jvm.JvmCrash
import io.appmetrica.analytics.impl.crash.service.ShouldSendCrashNowPredicate
import io.appmetrica.analytics.impl.utils.concurrency.ExclusiveMultiProcessFileLock
import io.appmetrica.analytics.impl.utils.concurrency.FileLocksHolder
import io.appmetrica.analytics.testutils.CommonTest
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verifyNoInteractions
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import java.io.File

@RunWith(RobolectricTestRunner::class)
internal class ReadAndReportRunnableTest : CommonTest() {

    private val crashFileName = "crash file name"

    private val crashFile: File = mock {
        on { name } doReturn crashFileName
        on { exists() } doReturn true
    }

    private val crash: JvmCrash = mock()

    private val fileReader: Function<File, JvmCrash?> = mock {
        on { apply(crashFile) } doReturn crash
    }

    private val finalizator: Consumer<File> = mock()
    private val crashConsumer: Consumer<JvmCrash> = mock()

    private val fileLock: ExclusiveMultiProcessFileLock = mock()

    private val fileLocksHolder: FileLocksHolder = mock {
        on { getOrCreate(crashFileName) } doReturn fileLock
    }

    private val shouldSendCrashPredicate: ShouldSendCrashNowPredicate<JvmCrash> = mock {
        on { shouldSend(crash) } doReturn true
    }

    private val readAndReportRunnable: ReadAndReportRunnable<JvmCrash> by setUp {
        ReadAndReportRunnable(
            crashFile,
            fileReader,
            finalizator,
            crashConsumer,
            fileLocksHolder,
            shouldSendCrashPredicate
        )
    }

    @Test
    fun run() {
        readAndReportRunnable.run()
        inOrder(fileLocksHolder, fileLock, crashConsumer, finalizator) {
            verify(fileLock).lock()
            verify(crashConsumer).consume(crash)
            verify(finalizator).consume(crashFile)
            verify(fileLock).unlockAndClear()
            verify(fileLocksHolder).clear(crashFileName)
        }
    }

    @Test
    fun `run if no crash file`() {
        whenever(crashFile.exists()).thenReturn(false)
        readAndReportRunnable.run()
        verifyNoInteractions(fileLocksHolder, fileLock, crashConsumer, finalizator)
    }

    @Test
    fun `run if crash is null`() {
        whenever(fileReader.apply(crashFile)).thenReturn(null)
        readAndReportRunnable.run()
        verifyNoInteractions(crashConsumer)
        inOrder(fileLock, finalizator, fileLocksHolder) {
            verify(fileLock).lock()
            verify(finalizator).consume(crashFile)
            verify(fileLock).unlockAndClear()
            verify(fileLocksHolder).clear(crashFileName)
        }
    }

    @Test
    fun `run if throw exception`() {
        whenever(fileReader.apply(crashFile)).thenThrow(RuntimeException())
        readAndReportRunnable.run()
        verifyNoInteractions(crashConsumer)
        inOrder(fileLock, finalizator, fileLocksHolder) {
            verify(fileLock).lock()
            verify(finalizator).consume(crashFile)
            verify(fileLock).unlockAndClear()
            verify(fileLocksHolder).clear(crashFileName)
        }
    }

    @Test
    fun `run if should not send crash predicate`() {
        whenever(shouldSendCrashPredicate.shouldSend(crash)).thenReturn(false)
        readAndReportRunnable.run()
        verifyNoInteractions(crashConsumer)
        inOrder(fileLock, finalizator, fileLocksHolder) {
            verify(fileLock).lock()
            verify(finalizator, never()).consume(crashFile)
            verify(fileLock).unlockAndClear()
            verify(fileLocksHolder).clear(crashFileName)
        }
    }
}
